keps/sig-api-machinery/365-paginated-lists/README.MD

```diff
@@ -742,3 +567,80 @@ the peak CPU and memory used inside the client, apiserver, or etcd processes.
 
 Various optimizations exist that can and should be applied to minimizing the amount of data that is transferred from
 etcd to the client or number of allocations made in each location, but do not change how response size scales with number of entries.
+
+## Potential future extensions
+
+Not all LIST requests are served from etcd. Our API allows clients to opt-in for the LIST result to be
+served from kube-apiserver cache (called watchcache), being propagated via watch from etcd.
+However, as of 1.28 release, watchcache still doesn't support pagination. This makes the logic that
+we use extremely non-intuitive, as it effectively works as following:
+
+- if the `continuation` is set - always pass the request to etcd (and honor the LIMIT)
+- if resourveVersion='0' - ignore the LIMIT parameter and return result from watchcache
+- if resourceVersion=<some concrete RV> and LIMIT is set - pass the request to etcd (and honor the LIMIT)
+- if resourceVersion=<some concrete RV> and LIMIT is not set - serve the request from watchcache
+   - which effectively results in different semantics about data freshness depending on LIMIT being set or not
+
+We propose to add support for pagination in watchcache as following:
+
+1. As of now, watchcache (in addition to transaction log which is not relevant for this proposal) contains
+  the current (consistent at a given RV) state of all objects of a given type. This is currently stored
+  in client-go `Indexer`, which internally stores that in hash-map.
+  We propose to replace the usage of `Indexer` with an implementation that will be build on top of
+  BTree that supports copy-on-write semantic. We are planning to use the implementation from
```

Comment:
Do we need to introduce a new data structure? It seems to me that the initial events are already being "copied" today. It appears that all we need to do is store this data in memory and add some helper fields for pagination.

Reply:
Won't work in HA setup, because the continuation could be send to a different kube-apiserver.

Reply:
In that case the servers could route requests between themselves, but I understand that it's not the solution we would like to implement :)

Reply:
What will it look like in the case of the proposed/described solution?
Will we be storing a copy of the BTree in the database?

Reply:
No need to store anything more in DB - we just need a btree (copy-on-write) in memory.
FWIW, a (not fully working) poc for that is in https://github.com/kubernetes/kubernetes/pull/108392
But this is out-of-scope for this KEP and if at all, will be proposed officially as a separate follow up KEP.